### PR TITLE
SoA implementation for the Axilrod-Teller-Muto Functor

### PR DESCRIPTION
**Pull request by thesamriel**
## SoA implementation for the Axilrod-Teller-Muto Functor

This PR adds the implementations for the SoAFunctors of the ATM functor and includes the following:
- [x] Renaming from AxilrodTeller.. to AxilrodTellerMuto.., which is more common in literature. (Edit: This was done separately in #1043 .
- [x] Implementation of `SoAFunctorSingle`, `SoAFunctorPair`, `SoAFunctorTriple`
- [x] Implementation of a Newton3 and non-Newton3 kernel
- [x] Correct calculation of globals and flop counting (includes a small correction in the virial calculation for AoS)
- [x] Tests for SoA functionality

Some remarks:
-  Tests for correctness of the computations have been added.
-  I don't use masking but 3 if-statements for the distance checks. Preliminary testing has shown that the early continuation saves more time than a possible autovectorization.
-  `SoAFunctorPair` and `SoAFunctorTriple` contain precalculations for those distances that would be computed multiple times during the triple loops. This proved beneficial during some local testing. 
- `SoAFunctorVerlet` is skipped for this PR and can be added once Verlet Lists algorithms are available
- Below graph shows preliminary comparison results on my local machine

image

## Update 19th December 2025

All 3 SoA functor methods have been optimized and use a specialized algorithm which is optimized for a cell size equal to the cutoff, which is assumed to be the most common case.

### `SoAFunctorSingle`

All pairwise distances are computed and stored in a flattened triangular matrix that serves as a lookup-table.
Then, we iterate over all particles in a basic triple for-loop.

### `SoAFunctorPair`

We iterate over all particles from one cell and build a 'neighbor list' of particles from the other cell. Then we find the triplets.
```
for Particle1 in SoA1
    for all particles in SoA2
         if (distance <= cutoff)
             ->Save particle and distance in NeighborList
    for Particle2 in NeighborList
        for Particle3 in NeighborList
            if (distance23 <= cutoff)
                computeForce

// Repeat the other way around
for Particle1 in SoA2
    for all particles in SoA1
        ...
```

### `SoAFunctorTriple`

We first find and save all valid pairs between SoA2 and SoA3. Then triplets are found for every pair and every particle from SoA1.
```
for Particle2 in SoA2
    for Particle3 in SoA3
        if (distance23 <= cutoff)
            ->Save Pair and Distance in Pairs23

for [Particle2, Particle3] in Pairs23
    for Particle1 in SoA1
        if (distance12 <= cutoff and distance13 <= cutoff)
            computeForce
```

### CODE REVIEW

**Comment by thesamriel** [comment]
@SamNewcome I squeezed in an additional commit for renaming with `git mv` (c4974aba7b77fd1f3e614bcb53a80453a853f68a), but it's still showing incorrectly for most files. 
It might be a GitHub issue? (see https://github.com/orgs/community/discussions/43177)
Not sure how I can do it otherwise.

**Comment by thesamriel** [comment]
Or hold on, I can do an intermediate PR for just the renaming first.

File: applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h [hunk 1]
```diff
@@ -288,6 +289,215 @@ class AxilrodTellerMutoFunctor
...
+        for (unsigned int k = j + 1; k < soaSize; ++k) {
+          const auto ownedStateK = ownedStatePtr[k];
+          if (ownedStateK == autopas::OwnershipState::dummy) {
+            continue;
+          }
+
+          SoAFloatPrecision nu = const_nu;
+          if constexpr (useMixing) {
+            nu = _PPLibrary->getMixingNu(typeptr[i], typeptr[j], typeptr[k]);
+          }
```
Comment (SamNewcome) [inline applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h]: Maybe it is faster to move this to after the "`continue`"

File: applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h [hunk 2]
```diff
@@ -497,20 +713,672 @@ class AxilrodTellerMutoFunctor
...
+      const SoAFloatPrecision &distXJK, const SoAFloatPrecision &distYJK, const SoAFloatPrecision &distZJK,
+      const SoAFloatPrecision &distXKI, const SoAFloatPrecision &distYKI, const SoAFloatPrecision &distZKI,
+      const SoAFloatPrecision &distSquaredIJ, const SoAFloatPrecision &distSquaredJK,
+      const SoAFloatPrecision &distSquaredKI, const SoAFloatPrecision &nu, SoAFloatPrecision &forceIX,
+      SoAFloatPrecision &forceIY, SoAFloatPrecision &forceIZ, SoAFloatPrecision &factor,
+      SoAFloatPrecision &allDotProducts, SoAFloatPrecision &allDistsSquared) const {
+    // Calculate prefactor
+    allDistsSquared = distSquaredIJ * distSquaredJK * distSquaredKI;
+    const SoAFloatPrecision allDistsTo5 = allDistsSquared * allDistsSquared * std::sqrt(allDistsSquared);
+    factor = 3.0 * nu / allDistsTo5;
```
Comment (SamNewcome) [inline applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h]: You could shift all your expensive square roots (and to a lesser extent, divisions) to where you do pairwise precomputations. Therefore, you only need to do a square root per pair as opposed to per triplet. You can then go one step further and for each pair calculate the inverse of each pairwise distance to the power of five. Then, here, you simply just need to multiply all three inverses of the pairwise distances to the power of five.

File: applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h [hunk 3]
```diff
@@ -135,6 +134,7 @@ class AxilrodTellerMutoFunctor
     }
     if constexpr (countFLOPs) {
       _aosThreadDataFLOPs.resize(autopas::autopas_get_max_threads());
+      AutoPasLog(DEBUG, "Using AxilrodTellerMutoFunctor with countFLOPs is not tested for SoA datalayout.");
```
Comment (SamNewcome) [inline applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h]: But it is now, right?

File: applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorFlopCounterTest.cpp [hunk 4]
```diff
@@ -143,69 +133,302 @@ void ATMFunctorFlopCounterTest::testFLOPCounterAoSOMP(bool newton3) {
...
+  }
+
+  constexpr double cutoff = 1.1;
+
+  // const auto particlePositions = getParticlePositions(functorFunctionType);
+  constexpr std::array p0 = {0.2, 0.2, 0.2};
+  constexpr std::array p1 = {1.0, 0.2, 0.2};
+  constexpr std::array p2 = {1.0, 0.8, 0.2};
+  constexpr std::array p3 = {0.2, 0.2, 2.5};
+  const std::vector<Molecule> molVec{Molecule(p0, {0, 0, 0}, 0), Molecule(p1, {0, 0, 0}, 1), Molecule(p2, {0, 0, 0}, 2),
```
Comment (SamNewcome) [inline applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorFlopCounterTest.cpp]:
```suggestion
  const std::vector molVec{Molecule(p0, {0, 0, 0}, 0), Molecule(p1, {0, 0, 0}, 1), Molecule(p2, {0, 0, 0}, 2),
```

File: applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorFlopCounterTest.cpp [hunk 5]
```diff
@@ -143,69 +133,302 @@ void ATMFunctorFlopCounterTest::testFLOPCounterAoSOMP(bool newton3) {
...
+      expectedNoN3GlobalsCalcs * numFLOPsPerNoN3GlobalsCall;
+
+  // We cannot predict the exact number of FLOPs as distance checks can be 1, 2 or 3 per triplet
+  ASSERT_LE(countedFlops, expectedFlopsMax) << "Tested for " << to_string(functorFunctionType);
+  ASSERT_GE(countedFlops, expectedFlopsMin) << "Tested for " << to_string(functorFunctionType);
+
+  const auto expectedHitRate =
+      (static_cast<double>(expectedN3KernelCalls) + static_cast<double>(expectedNoN3KernelCalls)) /
+      static_cast<double>(expectedNumTriplets);
+  ASSERT_NEAR(countedHitRate, expectedHitRate, 1e-14) << "Tested for " << to_string(functorFunctionType);
```
Comment (SamNewcome) [inline applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorFlopCounterTest.cpp]: Why `ASSERT` and not `EXPECT`?

File: applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorTestGlobals.cpp [hunk 6]
```diff
@@ -10,41 +10,24 @@
...
-    case allOutside:
-      whereFactor = 0.;
-      where_str = "outside";
-      owned1 = owned2 = owned3 = false;
-      break;
-    default:
-      FAIL() << "not in enum where_type";
-  }
+  // Map `where` to configuration values
+  const std::map<where_type, std::tuple<double, bool, bool, bool>> whereConfig = {
```
Comment (SamNewcome) [inline applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorTestGlobals.cpp]: Explain briefly what the tuple should represent

File: applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorTestGlobals.cpp [hunk 7]
```diff
@@ -73,15 +56,282 @@ void ATMFunctorTestGlobals<FuncType>::ATMFunctorTestGlobalsNoMixing(ATMFunctorTe
...
+  }
+
+  functor.endTraversal(newton3);
+
+  double potentialEnergy = functor.getPotentialEnergy();
+  double virial = functor.getVirial();
+
+  double expectedEnergy = calculateATMPotential(p1Pos, p2Pos, p3Pos, cutoff, nu);
+  auto [virial1, virial2, virial3] = calculateATMVirialTotalPerParticle(p1Pos, p2Pos, p3Pos, cutoff, nu);
+  double expectedVirial = virial1 * owned1 + virial2 * owned2 + virial3 * owned3;
```
Comment (SamNewcome) [inline applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorTestGlobals.cpp]: Make `const`

File: applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorTestGlobals.cpp [hunk 8]
```diff
@@ -73,15 +56,282 @@ void ATMFunctorTestGlobals<FuncType>::ATMFunctorTestGlobalsNoMixing(ATMFunctorTe
...
+
+/*
+ * Tests potential energy and virial calculations for a triplet split over periodic boundary conditions.
+ * This especially interesting for the virial
+ * @tparam FuncType The type of the functor to test
+ * @param functorToTest Test either AoS, SoAPair or SoATriple
+ * @param newton3 Whether to use newton 3
+ */
+template <class FuncType>
+void ATMFunctorTestGlobals<FuncType>::ATMFunctorTestGlobalsPeriodicBCs(std::string functorToTest, bool newton3) {
```
Comment (SamNewcome) [inline applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorTestGlobals.cpp]:
```suggestion
void ATMFunctorTestGlobals<FuncType>::ATMFunctorTestGlobalsPeriodicBCs(const std::string& functorToTest, bool newton3) {
```

File: applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorTestNoGlobals.cpp [hunk 9]
```diff
@@ -223,325 +237,200 @@ TYPED_TEST_P(ATMFunctorTestNoGlobals, testSoANoGlobalsATM) {
...
+    if (newton3 or soaFunctorType == TestType::SoAFunctorType::single or
+        soaFunctorType == TestType::SoAFunctorType::verlet) {
+      const auto expectedForceIter1P3 = mixing ? this->expectedForceMixingP3 : this->expectedForceNonMixingP3;
+      testNonZeroForce(*f3, expectedForceIter1P3, ATMFunctorTest::to_string(soaFunctorType));
     } else {
-      EXPECT_DOUBLE_EQ(f3[0], 0);
-      EXPECT_DOUBLE_EQ(f3[1], 0);
-      EXPECT_DOUBLE_EQ(f3[2], 0);
+      testZeroForce(*f3, ATMFunctorTest::to_string(soaFunctorType));
     }
```
Comment (SamNewcome) [inline applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorTestNoGlobals.cpp]: I think this way of looping through the three particles is quite messy. Maybe, you could simply create an array/vector of particles, and simply loop through each cell, adding particles to the array, or skipping the cell if empty. Then apply the tests. Alternatively, simply write the whole thing with a type-case switch - I feel the advantages gained from reduced code duplication you have here are lost from having all these complex `if`-statements and branching. 

In any case, also remove the pointers - you can just use references.

File: applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorTestNoGlobals.cpp [hunk 10]
```diff
@@ -223,325 +237,200 @@ TYPED_TEST_P(ATMFunctorTestNoGlobals, testSoANoGlobalsATM) {
...
+          f3 = &(++cell2.begin())->getF();
           break;
-        case TestType::InteractionType::triple:
-          f3 = cell3.begin()->getF();
+        case TestType::SoAFunctorType::triple:
+          f3 = &cell3.begin()->getF();
           break;
         default:
           break;
       }
```
Comment (SamNewcome) [inline applicationLibrary/molecularDynamics/tests/singleSiteTests/ATMFunctorTestNoGlobals.cpp]: Similarly to above

File: src/autopas/tuning/AutoTuner.cpp [hunk 11]
```diff
@@ -362,15 +362,25 @@ size_t AutoTuner::getCurrentNumSamples() const {
...
-  // if there is no data for the non rebuild iterations we have to assume them taking the same time as rebuilding ones
-  // this might neither be a good estimate nor fair but the best we can do
-  const auto reducedValueNotBuilding =
-      _samplesNotRebuildingNeighborLists.empty()
-          ? reducedValueBuilding
-          : autopas::OptimumSelector::optimumValue(_samplesNotRebuildingNeighborLists, _selectorStrategy);
+  bool hasRebuildSamples = !_samplesRebuildingNeighborLists.empty();
+  bool hasNotRebuildSamples = !_samplesNotRebuildingNeighborLists.empty();
+
+  if (!hasRebuildSamples && !hasNotRebuildSamples) {
```
Comment (SamNewcome) [inline src/autopas/tuning/AutoTuner.cpp]:
```suggestion
  if (not hasRebuildSamples and not hasNotRebuildSamples) {
```

**Review by SamNewcome — CHANGES_REQUESTED** [review]
Some refactoring for clarity and some suggestions to move expensive square root & division operations out of the triwise kernel and into the pairwise pre-computations.

File: src/autopas/LogicHandler.h [hunk 12]
```diff
@@ -1252,7 +1252,10 @@ IterationMeasurements LogicHandler<Particle_T>::computeInteractions(Functor &fun
   auto &autoTuner = *_autoTunerRefs[interactionType];
 #ifdef AUTOPAS_ENABLE_DYNAMIC_CONTAINERS
   if (autoTuner.inFirstTuningIteration()) {
-    autoTuner.setRebuildFrequency(getMeanRebuildFrequency(/* considerOnlyLastNonTuningPhase */ true));
+    // Don't allow a rebuild frequency of 0
+    if (auto meanRebuildFrequency = getMeanRebuildFrequency(/* considerOnlyLastNonTuningPhase */ true)) {
```
[the remark this answers is not in the document]
Comment (thesamriel) [reply src/autopas/LogicHandler.h]: I don't quite get what you mean. But using your suggested changes in [this comment](https://github.com/AutoPas/AutoPas/pull/1039/files/021c4189a24ebc26f0dc0caf59affe200caa7605#r2505013336) for the critical `AutoTuner` method, it's fine again to have a 0 rebuild frequency, so I reverted this change in [6a79246](https://github.com/AutoPas/AutoPas/pull/1039/commits/6a7924646d6bb3e3b08263b486f91d412db6c205)

File: applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h (hunk 2 above)
Comment (SamNewcome) [reply applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h]: I think these should be fairly simple & quick to implement. I'll see if I get time tomorrow.

File: applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h [hunk 13]
```diff
@@ -497,20 +712,672 @@ class AxilrodTellerMutoFunctor
...
+      // CASE: Particle i and j are both in soa1, k is in soa2
+      for (unsigned int j = i + 1; j < soa1.size(); ++j) {
+        const auto ownedStateJ = ownedStatePtr1[j];
+        if (ownedStateJ == autopas::OwnershipState::dummy) {
+          continue;
+        }
+
+        SoAFloatPrecision fXAccJ = 0.;
+        SoAFloatPrecision fYAccJ = 0.;
+        SoAFloatPrecision fZAccJ = 0.;
```
Comment (dmartin13) [inline applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h]: While vectorizing the SoAFunctorPair, I noticed that this accumulator is initialized and later stored, but never used. Instead, the k-loop always writes directly to the memory of particle j. I think this accumulator could be used to save memory accesses. I also think an accumulator could be used for j in the first loop (// i from soa1, j and k from soa2).

This is the general structure at the moment:
```
for (i...) {
  AccI = 0
  // i from soa1, j and k from soa2
  for (j ...) {
    for (k ...) {
      if (noN3) {
        AccI += forceI
      } else {
        AccI += forceI
        store(forceJ)
        store(forceK)
      }
    }
  }
  // i anf j from soa1, k from soa2
  for (j ...) {
    AccJ = 0
    for (k...) {
      AccI += forceI
      store(forceJ)
      if(N3) {
        store(forceK)
      }
    }
    store(AccJ)
  }
  store(AccI)
}
```
That's how I'm using it currently in the vectorized functor, and I think it could also be adapted here:
```
for (i...) {
  AccI = 0
  // i from soa1, j and k from soa2
  for (j ...) {
    AccJ = 0
    for (k ...) {
      if (noN3) {
        AccI += forceI
      } else {
        AccI += forceJ
        AccJ += forceJ
        store(forceK)
      }
    }
    store(AccJ)
  }

  for (j ...) {
    AccJ = 0
    for (k...) {
      AccI += forceI
      AccJ += forceJ
      if(N3) {
        store(forceK)
      }
    }
    store(AccJ)
  }
  store(AccI)
}
```

File: applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h [hunk 14]
```diff
@@ -288,6 +288,215 @@ class AxilrodTellerMutoFunctor
...
+    size_t numTripletsCountingSum = 0;
+    size_t numDistanceCalculationSum = 0;
+    size_t numKernelCallsN3Sum = 0;
+    size_t numGlobalCalcsSum = 0;
+
+    const SoAFloatPrecision const_nu = _nu;
+    const size_t soaSize = soa.size();
+    if constexpr (countFLOPs) {
+      numTripletsCountingSum = soaSize * (soaSize - 1) * (soaSize - 2) / 6;
+    }
```
Comment (SamNewcome) [inline applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h]:
```suggestion
    const size_t soaSize = soa.size();
    const size_t numTriplets = countFLOPs ? soaSize * (soaSize - 1) * (soaSize - 2) / 6 : 0; // Only needed if counting FLOPs
    size_t numDistanceCalculationSum = 0;
    size_t numKernelCallsN3Sum = 0;
    size_t numGlobalCalcsSum = 0;

    const SoAFloatPrecision const_nu = _nu;
```
Only spotted this when reviewing David's branch. -> Clearer name & using `const` makes it clearer that this isn't an accumulated sum.

Comment (thesamriel) [reply applicationLibrary/molecularDynamics/molecularDynamicsLibrary/AxilrodTellerMutoFunctor.h]: Good point! I actually moved it to the end of the functor methods, as it is anyways needed only for the accumulation in the end. (8951de7)

**Comment by thesamriel** [comment]
@SamNewcome I went over all of your suggestions that might improve performance and in the end, ended up with even bigger changes 😅.
 All Functors have been changed. I explained each algorithm briefly in the updated description of this PR.
 I will add a performance comparison between the new and the original proposed implementation.
 
 I did not have any success in pulling out the square root and divisions from the kernel.
